base/applications/certutil/certutil.c

```diff
@@ -0,0 +1,149 @@
+        {
+            BYTE Buffer[2048];
+            DWORD cbRead;
+
+            while ((bSuccess = ReadFile(hFile, Buffer, sizeof(Buffer), &cbRead, NULL)))
+            {
+                if (cbRead == 0)
+                    break;
+
+                if (!CryptHashData(hHash, Buffer, cbRead, 0))
+                {
+                    bSuccess = FALSE;
+                    ConPrintf(StdOut, L"CertUtil: -hashfile command failed to hash: %d\n", GetLastError());
+                }
+            }
+
+            if (bSuccess)
+            {
+                BYTE rgbHash[20];
+                DWORD cbHash, n;
+
+                if (CryptGetHashParam(hHash, HP_HASHVAL, rgbHash, &cbHash, 0))
+                {
+                    ConPrintf(StdOut, L"SHA1 hash of %s:\n", Filename);
+                    for (n = 0; n < cbHash; ++n)
+                    {
+                        ConPrintf(StdOut, L"%02x", rgbHash[n]);
+                    }
+                    ConPuts(StdOut, L"\n");
+                }
+                else
+                {
+                    ConPrintf(StdOut, L"CertUtil: -hashfile command failed to extract hash: %d\n", GetLastError());
+                    bSuccess = FALSE;
+                }
+            }
+
+            CryptDestroyHash(hHash);
+        }
+        else
+        {
+            ConPrintf(StdOut, L"CertUtil: -hashfile command no algorithm: %d\n", GetLastError());
+        }
+
+        CryptReleaseContext(hProv, 0);
+    }
+    else
+    {
+        ConPrintf(StdOut, L"CertUtil: -hashfile command no context: %d\n", GetLastError());
+    }
+
+    CloseHandle(hFile);
+    return bSuccess;
+}
+
+
+static void print_usage()
+{
+    ConPuts(StdOut, L"Verbs:\n");
+    ConPuts(StdOut, L"  -hashfile           -- Display cryptographic hash over a file\n");
+    ConPuts(StdOut, L"\n");
+    ConPuts(StdOut, L"CertUtil -?           -- Display a list of all verbs\n");
+    ConPuts(StdOut, L"CertUtil -hashfile -? -- Display help text for the 'hashfile' verb\n");
+}
+
+int wmain(int argc, WCHAR *argv[])
+{
+    int n;
+
+    /* Initialize the Console Standard Streams */
+    ConInitStdStreams();
+
+    if (argc == 1) /* i.e. no commandline arguments given */
+    {
+        print_usage();
+        return EXIT_SUCCESS;
+    }
+
+    for (n = 1; n < argc; ++n)
+    {
+        if (!_wcsicmp(argv[n], L"-?"))
+        {
+            print_usage();
+            return EXIT_SUCCESS;
+        }
+        else if (!_wcsicmp(argv[n], L"-hashfile"))
+        {
+            if (argc == 3)
+            {
+                if (!_wcsicmp(argv[n+1], L"-?"))
+                {
+                    print_usage();
+                    return EXIT_SUCCESS;
+                }
+                else
+                {
+                    BOOL bSuccess = hash_file(argv[n+1]);
+                    ConPrintf(StdOut, L"CertUtil: -hashfile command %s\n", bSuccess ? L"completed successfully" : L"FAILED");
```

Comment:
You're already printing the failure message in all failure cases, so you'll get double
```
>base\applications\certutil\certutil.exe -hashfile C:\bootnx
CertUtil: -hashfile command failed: 2
CertUtil: -hashfile command FAILED
```

Reply:
yes, following windows output (more or less)